### PR TITLE
Update step-40.

### PR DESCRIPTION
**Pull request by drwells**
Part of #6774.

### CODE REVIEW

File: examples/step-40/step-40.cc [hunk 1]
```diff
@@ -693,17 +690,27 @@ namespace Step40
...
+// also important: Utilities::MPI::MPI_InitFinalize::~MPI_InitFinalize() calls
+// <code>PetscFinalize</code> (and finalization functions for other
+// libraries), which will delete any in-use PETSc objects. This must be done
+// after we destruct the Laplace solver to avoid double deletion
+// errors. Fortunately, due to the order of destructor call rules of C++, we
+// do not need to worry about any of this: everything happens in the correct
+// order (i.e., the reverse of the order of construction). The last function
+// called by Utilities::MPI::MPI_InitFinalize::~MPI_InitFinalize() is
+// <code>MPI_Finalize</code>: i.e., once this object is destructed the program
+// should exit since MPI will no longer be available.
```
Comment (drwells) [inline examples/step-40/step-40.cc]: I couldn't figure out what the original text meant by 'pair of braces' so I just rewrote the whole thing.

File: examples/step-40/step-40.cc [hunk 2]
```diff
@@ -403,19 +400,18 @@ namespace Step40
...
-                                          fe_values.JxW(q_point));
+                    cell_matrix(i, j) += fe_values.shape_grad(i, q_point) *
+                                         fe_values.shape_grad(j, q_point) *
+                                         fe_values.JxW(q_point);
 
-                  cell_rhs(i) +=
-                    (rhs_value * fe_values.shape_value(i, q_point) *
-                     fe_values.JxW(q_point));
+                  cell_rhs(i) += rhs_value * fe_values.shape_value(i, q_point) *
+                                 fe_values.JxW(q_point);
```
Comment (masterleinad) [inline examples/step-40/step-40.cc]: Should we split this into three lines?

File: examples/step-40/step-40.cc (hunk 1 above)
Comment (masterleinad) [reply examples/step-40/step-40.cc]: Sounds good!

File: examples/step-40/step-40.cc [hunk 3]
```diff
@@ -280,13 +280,13 @@ namespace Step40
...
     // As with all other things in %parallel, the mantra must be that no
     // processor can store all information about the entire universe. As a
-    // consequence, we need to tell the constraints object for which degrees
-    // of freedom it can store constraints and for which it may not expect any
-    // information to store. In our case, as explained in the @ref distributed
-    // module, the degrees of freedom we need to care about on each processor
-    // are the locally relevant ones, so we pass this to the
-    // ConstraintMatrix::reinit function. As a side note, if you forget to
-    // pass this argument, the ConstraintMatrix class will allocate an array
+    // consequence, we need to tell the affine constraints object for which
```
[the remark this answers is not in the document]
Comment (drwells) [reply examples/step-40/step-40.cc]: I agree.

File: examples/step-40/step-40.cc [hunk 4]
```diff
@@ -280,13 +280,13 @@ namespace Step40
...
+    // degrees of freedom it can store constraints and for which it may not
+    // expect any information to store. In our case, as explained in the @ref
+    // distributed module, the degrees of freedom we need to care about on
```
[the remark this answers is not in the document]
Comment (drwells) [reply examples/step-40/step-40.cc]: It took me a few minutes to figure out what you meant. I fixed this (and checked the other ones).

File: examples/step-40/step-40.cc (hunk 2 above)
Comment (drwells) [reply examples/step-40/step-40.cc]: We do that everywhere else so yes, lets do it here too.

**Comment by drwells** [comment]
I fixed everything that was commented on. I also added some more text that clarifies that one can use either PETSc or Trilinos (and, subsequently, either Hypre or ML).

File: examples/step-40/step-40.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply examples/step-40/step-40.cc]: I think you need to put the `@ref` onto the same line as the tag.

[the remark this answers is not in the document]
Comment (drwells) [reply examples/step-40/step-40.cc]: @bangerth I am not sure that I understand your comment. I think I fixed all of them in a subsequent commit: am I missing one?

[the remark this answers is not in the document]
Comment (masterleinad) [reply examples/step-40/step-40.cc]: I believe he was looking at an outdated version.

[the remark this answers is not in the document]
Comment (drwells) [reply examples/step-40/step-40.cc]: That seems likely :)

[the remark this answers is not in the document]
Comment (bangerth) [reply examples/step-40/step-40.cc]: Ah, correct, yes. Always behind the times these days -- reviewing and merging dozens of ASPECT pull requests a day distracts a lot from looking at deal.II stuff :-)

**Pull request merged**